lib/dai-plugin-mcd/src/ManagedCdp.js

```diff
@@ -12,24 +12,88 @@ export default class ManagedCdp {
 
     assert(ilk && typeof ilk === 'string', 'Must specify ilk');
     this.ilk = ilk;
-
     this._cdpManager = cdpManager;
-    this.currency = cdpManager
-      .get(ServiceRoles.CDP_TYPE)
-      .getCdpType(null, ilk).currency;
+    this.type = cdpManager.get(ServiceRoles.CDP_TYPE).getCdpType(null, ilk);
+    this.currency = this.type.currency;
+    this._getUrnInfo();
+    this._getIlkInfo();
+  }
+
+  async getCollateralAmount() {
+    return this.currency.wei((await this._getUrnInfo()).ink);
   }
 
   async getCollateralValue() {
-    return this.currency.wei((await this._urnInfo()).ink);
+    const [collateral, price] = await Promise.all([
+      this.getCollateralAmount(),
+      this.type.getPrice()
+    ]);
+    return collateral.times(price);
   }
 
   async getDebtValue() {
-    const cdpType = this._cdpManager
-      .get(ServiceRoles.CDP_TYPE)
-      .getCdpType(this.currency, this.ilk);
-    const { rate } = await cdpType.ilkInfo();
-    const art = MDAI.wei((await this._urnInfo()).art);
-    return art.times(rate).shiftedBy(-27);
+    const [ilkInfo, urnInfo] = await Promise.all([
+      this._getIlkInfo(),
+      this._getUrnInfo()
+    ]);
+    const art = MDAI.wei(urnInfo.art);
+    return art.times(ilkInfo.rate).shiftedBy(-27);
+  }
+
+  async getCollateralizationRatio() {
+    const [collateral, debt] = await Promise.all([
+      this.getCollateralValue(),
+      this.getDebtValue()
+    ]);
+    assert(debt.gt(0), 'Debt cannot be zero');
+    return collateral.div(debt);
+  }
+
+  async getLiquidationPrice() {
+    const [liquidationRatio, debt, collateral] = await Promise.all([
+      this.type.getLiquidationRatio(),
+      this.getDebtValue(),
+      this.getCollateralAmount()
+    ]);
+    assert(collateral.gt(0), 'Collateral cannot be zero');
+    return debt.times(liquidationRatio).div(collateral);
+  }
+
+  async isSafe() {
+    const [liqPrice, collPrice] = await Promise.all([
+      this.getLiquidationPrice(),
+      this.type.getPrice()
+    ]);
+    return collPrice.gte(liqPrice);
+  }
+
+  async minCollateral() {
+    const [debt, liquidationRatio, price, collateral] = await Promise.all([
+      this.getDebtValue(),
+      this.type.getLiquidationRatio(),
+      this.type.getPrice(),
+      this.getCollateralAmount()
+    ]);
+    const lockedCollateral = debt.times(liquidationRatio).div(price);
```

Comment:
please don't use the word `locked` here since it already means something else in Maker lingo -- `minSafe` would work